src/Compilers/Core/Portable/Text/StringTextWriter.cs

```diff
@@ -33,9 +33,13 @@ public override Encoding Encoding
             get { return _encoding!; }
         }
 
-        public override SourceText ToSourceText()
+        public override SourceText ToSourceTextAndClear()
```

Comment:
I have a couple of thoughts on this:

1. We can definitely never, ever, expose this publicly. I'm somewhat shocked to find we have a `StringBuilderText` at all, tbh, since that's using a mutable backing store for a `SourceText`.
2. Do we have any measurements on whether we're going to increase CPU load during later indexing by not calling `ToString()` here? IE, are we trading memory for CPU time today?

Reply:
1) I definitely agree that StringBuilderText should never be publicly exposed. I renamed the ToSourceText method to be more explicit about it's behavior because of this concern.

2) I didn't take any benchmarks with this change. I suppose future calls of ToString when passed the full span on the StringBuilderText would be more expensive than on a StringText, but other than that, I wouldn't imagine the standard indexing or CopyTo calls would be decidedly different wrt performance.

Reply:
I went ahead and created some benchmarks of the form below,. The thought here was to measure the combined cost for creation and a single access at every position. I had thought that the allocation improvement would more than pay for whatever additional cost indexing into a stringbuilder has over indexing into a string. However, that doesn't appear to be the case.

In summary, this looks to be about 10% slower (although half the allocation cost) so I'll go ahead and kill this part of the PR (will keep the changes in CompositeText as I think those are still valuable). 

```
| Method |  Size |       Mean |     Error |    StdDev |      Gen 0 |    Gen 1 | Gen 2 | Allocated |
|------- |------ |-----------:|----------:|----------:|-----------:|---------:|------:|----------:|
|    Old |  1000 |   3.031 ms | 0.0240 ms | 0.0200 ms |   980.4688 |        - |     - |      4 MB |
|    New |  1000 |   3.539 ms | 0.0672 ms | 0.0629 ms |   531.2500 |        - |     - |      2 MB |
|    Old | 10000 |  30.039 ms | 0.2080 ms | 0.1844 ms |  9281.2500 |  31.2500 |     - |     38 MB |
|    New | 10000 |  33.741 ms | 0.3000 ms | 0.2806 ms |  4666.6667 |  66.6667 |     - |     19 MB |
|    Old | 40000 | 113.235 ms | 1.8486 ms | 1.7292 ms | 37000.0000 | 200.0000 |     - |    153 MB |
|    New | 40000 | 124.594 ms | 0.8420 ms | 0.7031 ms | 18500.0000 | 250.0000 |     - |     77 MB |
```
------------
```
        [Benchmark]
        public void Old()
        {
            var content = new string('a', Size);

            for (var loopIndex = 0; loopIndex < _loopCount; loopIndex++)
            {
                var writer = SourceTextWriter.CreateOld(Encoding.UTF8, SourceHashAlgorithm.Sha256, Size);
                writer.Write(content);

                var sourceText = writer.ToSourceTextAndClear();

                for (var i = 0; i < sourceText.Length; i++)
                {
                    var curChar = sourceText[i];
                }
            }
        }
```

Reply:
@333fred -- Thoughts on this without the switch to using StringBuilderText?